pkg/portForward/portForward.go

```diff
@@ -0,0 +1,148 @@
+package portForward
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/library/pkg/devfile/parser"
+	parsercommon "github.com/devfile/library/pkg/devfile/parser/data/v2/common"
+
+	"github.com/redhat-developer/odo/pkg/kclient"
+	"github.com/redhat-developer/odo/pkg/libdevfile"
+	"github.com/redhat-developer/odo/pkg/log"
+	"github.com/redhat-developer/odo/pkg/state"
+	"github.com/redhat-developer/odo/pkg/util"
+)
+
+var _ Client = (*PFClient)(nil)
+
+type PFClient struct {
+	kubernetesClient kclient.ClientInterface
+	stateClient      state.Client
+
+	appliedEndpoints map[string][]int
+
+	// stopChan on which to write to stop the port forwarding
+	stopChan chan struct{}
+	// finishedChan is written when the port forwarding is finished
+	finishedChan chan struct{}
+}
+
+func NewPFClient(kubernetesClient kclient.ClientInterface, stateClient state.Client) *PFClient {
+	return &PFClient{
+		kubernetesClient: kubernetesClient,
+		stateClient:      stateClient,
+	}
+}
+
+func (o *PFClient) StartPortForwarding(
+	devFileObj parser.DevfileObj,
+	randomPorts bool,
+	errOut io.Writer,
+) error {
+
+	// get the endpoint/port information for containers in devfile and setup port-forwarding
+	containers, err := devFileObj.Data.GetComponents(parsercommon.DevfileOptions{
+		ComponentOptions: parsercommon.ComponentOptions{ComponentType: v1alpha2.ContainerComponentType},
+	})
+	if err != nil {
+		return err
+	}
+	ceMapping := libdevfile.GetContainerEndpointMapping(containers)
+	if o.stopChan != nil && reflect.DeepEqual(ceMapping, o.appliedEndpoints) {
+		return nil
+	}
+
+	o.appliedEndpoints = ceMapping
+
+	o.StopPortForwarding()
+
+	o.stopChan = make(chan struct{}, 1)
+	o.finishedChan = make(chan struct{}, 1)
+
+	var portPairs map[string][]string
+	if randomPorts {
+		portPairs = randomPortPairsFromContainerEndpoints(ceMapping)
+	} else {
+		portPairs = portPairsFromContainerEndpoints(ceMapping)
+	}
+	var portPairsSlice []string
+	for _, v1 := range portPairs {
+		portPairsSlice = append(portPairsSlice, v1...)
+	}
+	pod, err := o.kubernetesClient.GetPodUsingComponentName(devFileObj.GetMetadataName())
+	if err != nil {
+		return err
+	}
+
+	// Output that the application is running, and then show the port-forwarding information
+	log.Info("\nYour application is now running on the cluster")
+
+	portsBuf := NewPortWriter(log.GetStdout(), len(portPairsSlice), ceMapping)
+	go func() {
+		err = o.kubernetesClient.SetupPortForwarding(pod, portPairsSlice, portsBuf, errOut, o.stopChan)
+		if err != nil {
+			fmt.Printf("failed to setup port-forwarding: %v\n", err)
+		}
+		o.finishedChan <- struct{}{}
```

Comment:
What is this for? When the user hits `Ctrl+C`, this channel will get populated? But if one calls `StopPortForwarding` method instead of hitting Ctrl+C, doesn't the `<-o.finishedChan` become a blocking call?

Reply:
the stopChan and finishedChan are kind of signals to communicate between the main go routine aand the go routine running the port forward.

The main go routine sends a "stop" signal to the port forward go routine (which is blocked at this `select`: https://github.com/redhat-developer/odo/blob/dce95d62e24d2ba18306f685e9a092f503d77b38/vendor/k8s.io/client-go/tools/portforward/portforward.go#L230)

When the PF go routine receives the signal, it will take some time to stop the port forwarding (on this line: https://github.com/redhat-developer/odo/blob/dce95d62e24d2ba18306f685e9a092f503d77b38/vendor/k8s.io/client-go/tools/portforward/portforward.go#L189),and after doing this, we can send a "finished" signal back to main go routine to indicate that the port forwarding is stopped, and so it is safe to recreate port forwardings.